### PR TITLE
4x ENS functionality

### PR DESCRIPTION
**Pull request by jdevcs**
## Description

https://github.com/ChainSafe/web3.js/issues/4552
https://github.com/ChainSafe/web3.js/issues/4602
https://github.com/ChainSafe/web3.js/issues/4553

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](../CONTRIBUTIONS.md).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

File: packages/web3-eth-ens/src/resolver.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		resolverContract: Contract<typeof RESOLVER>,
+		methodName: string,
+	) {
+		if (interfaceIds[methodName] === undefined) throw new Error(); //  ResolverMethodMissingError(resolver address, methodName);
+
+		let supported = false;
+		try {
+			supported = (await resolverContract.methods
+				.supportsInterface(interfaceIds[methodName])
+				.call()) as Awaited<Promise<boolean>>;
```
Comment (nazarhussain) [inline packages/web3-eth-ens/src/resolver.ts]: If contract typing is correct, then we don't need to use `as` here. Please check what type error you were getting here.

File: packages/web3-eth-ens/src/resolver.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		} catch (err) {
+			//  TODO throw new Error
+		}
+
+		if (!supported) throw new Error(); // new ResolverMethodMissingError(resolver address, methodName);
```
Comment (nazarhussain) [inline packages/web3-eth-ens/src/resolver.ts]:
```suggestion
		if (!supported) throw new Error(); // new ResolverMethodMissingError(resolver address, methodName);
		
		return supported; 
```

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: packages/web3-eth-ens/src/resolver.ts (hunk 1 above)
Comment (jdevcs) [reply packages/web3-eth-ens/src/resolver.ts]: `error  Unsafe assignment of an any value  @typescript-eslint/no-unsafe-assignment`

File: packages/web3-eth-ens/src/resolver.ts (hunk 2 above)
Comment (jdevcs) [reply packages/web3-eth-ens/src/resolver.ts]: I wasnt using boolean as return result

File: packages/web3-eth-ens/src/config.ts [hunk 3]
```diff
@@ -1 +1,22 @@
...
+
+// https://docs.ens.domains/contract-developer-guide/writing-a-resolver
+// resolver interface Ids
+export const interfaceIds: { [T: string]: string } = {
+	addr: '0x3b3b57de',
+	name: '0x691f3431',
+	abi: '0x2203ab56',
+	pubkey: '0xc8690233',
+	text: '0x59d1d43c',
+	contenthash: '0xbc1c58d1',
```
Comment (luu-alex) [inline packages/web3-eth-ens/src/config.ts:11]: in the docs it suggests to support `supportsInterface` with the interface id of `0x01ffc9a7`

File: packages/web3-eth-ens/src/ens.ts [hunk 4]
```diff
@@ -116,46 +119,79 @@ export class ENS {
...
+		return this._resolver.getContenthash(ENSName);
+	}
 
 	// TODO after eth.net.getNetworkType is complete
 	// public checkNetwork (): boolean {
 	// return true;
 	//  };
-
-	// TODO finish in resolver
-	// public supportsInterface(){
```
Comment (luu-alex) [inline packages/web3-eth-ens/src/ens.ts:158]: ENS should also include the supportsinterface method from resolver

**Comment by luu-alex** [comment]
i think unit tests for the resolver class would be good, other then that LGTM

File: packages/web3-eth-ens/src/resolver.ts (hunk 1 above)
Comment (nazarhussain) [reply packages/web3-eth-ens/src/resolver.ts]: @jdevcs  That's not right. Please create an issue to address this later.

**Review by nazarhussain — APPROVED** [review]

File: packages/web3-eth-ens/src/config.ts (hunk 3 above)
Comment (jdevcs) [reply packages/web3-eth-ens/src/config.ts:11]: Internally we are not sending any eth_call to `supportsInterface` for checking if resolver contract have `supportsInterface` . For external calls, If user want he can directly check it via [this](https://github.com/ChainSafe/web3.js/blob/ad2ee10234ba1f0c96d1dca48bb1c0d9bb4769cf/packages/web3-eth-ens/src/resolver.ts#L75)

**Comment by jdevcs** [comment]
> i think unit tests for the resolver class would be good, other then that LGTM

@luu-alex Resolver class is only sending transactions and sending eth_calls for reads so I think we can test it in integration tests.

File: packages/web3-eth-ens/src/resolver.ts (hunk 1 above)
Comment (jdevcs) [reply packages/web3-eth-ens/src/resolver.ts]: I was also wondering with typed contract feature it should have correct type.

**Review by luu-alex — APPROVED** [review]

File: packages/web3-eth-ens/src/resolver.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	private async getResolverContractAdapter(ENSName: string) {
+		//  TODO : (Future 4.1.0 TDB) cache resolver contract if frequently queried same ENS name, refresh cache based on TTL and usage, also limit cache size, optional cache with a flag
+		return this.registry.getResolver(ENSName);
+	}
+
+	//  https://eips.ethereum.org/EIPS/eip-165
+	// eslint-disable-next-line class-methods-use-this
+	public async checkInterfaceSupport(
```
Comment (Spacesai1or) [inline packages/web3-eth-ens/src/resolver.ts:26]: This could be a static method?

**Review by Spacesai1or — COMMENTED** [review]
You could do this in a subsequent PR, but I think `packages/web3-eth-ens/src/ens.ts` should have sanity check unit tests that test `this._resolver` is receiving the expected arguments from `ENS` class - this ensures the user facing API never changes accidentally

File: packages/web3-eth-ens/src/resolver.ts (hunk 5 above)
Comment (luu-alex) [reply packages/web3-eth-ens/src/resolver.ts:26]: I dont think so, it calls the resolverContract on line 38-40

Comment (luu-alex) [reply packages/web3-eth-ens/src/resolver.ts:26]: I dont think this matters too much, resolver is only used internally by us and is only exposed through the ENS class

**Comment by luu-alex** [comment]
@spacesailor24 i included add sanity check for resolver in the ens epic, so we can merge this one in

**Pull request merged**